modules/broker-core/src/main/java/io/ballerina/messaging/broker/core/transaction/Registry.java

```diff
@@ -81,25 +100,45 @@ public synchronized void prepare(Xid xid) throws ValidationException, BrokerExce
         branch.setState(Branch.State.PREPARED);
     }
 
+    private void checkForBranchExpiration(Branch branch) throws ValidationException {
+        if (branch.isExpired() || !cancelTimeoutTask(branch)) {
+            unregister(branch.getXid());
+            throw new ValidationException(TIMED_OUT_ERROR_MSG + branch.getXid());
+        }
+    }
+
     public synchronized void commit(Xid xid, boolean onePhase) throws ValidationException, BrokerException {
         Branch branch = branchMap.get(xid);
         validateCommitRequest(xid, onePhase, branch);
+
         branch.clearAssociations();
         branch.setState(Branch.State.FORGOTTEN);
         branch.commit(onePhase);
         unregister(xid);
     }
 
+    private boolean cancelTimeoutTask(Branch branch) {
+        Future timeoutTaskFuture = branch.getTimeoutTaskFuture();
+        return Objects.isNull(timeoutTaskFuture)
+                || timeoutTaskFuture.cancel(false)
+                || timeoutTaskFuture.isCancelled();
```

Comment:
Shouldn't we first check isCalled and then cancel if false rather than the other way?

Reply:
fixed!